Error with the PackageUpdateExternalAnnotator

Hi there! Thanks for making this plugin!

## What happened

I've opened the IDE and the pubspec.yaml file was opened

## Version information

IntelliJ IDEA `2023.2.3` • Flutter plugin `io.flutter 75.1.4` • Dart plugin `232.9559.10`Flutter 3.13.7 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 2f708eb839 (5 days ago) • 2023-10-09 09:58:08 -0500
Engine • revision a794cf2681
Tools • Dart 3.1.3 • DevTools 2.25.0

## Exception

*** exception class was changed or removed

```
ExternalToolPass: 
   
   com.intellij.diagnostic.PluginException: annotator: de.mariushoefler.flutterenhancementsuite.inspections.PackageUpdateExternalAnnotator@3d66aacf (class de.mariushoefler.flutterenhancementsuite.inspections.PackageUpdateExternalAnnotator) [Plugin: de.mariushoefler.flutter_enhancement_suite]
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:23)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:89)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.processError(ExternalToolPass.java:251)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:210)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:201)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.lambda$run$0(ExternalToolPass.java:164)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.runChangeAware(ExternalToolPass.java:266)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.lambda$run$2(ExternalToolPass.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:61)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.runUnderDisposeAwareIndicator(BackgroundTaskUtil.java:360)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:162)
	at com.intellij.util.ui.update.Update.runUpdate(Update.java:114)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:348)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:338)
	at com.intellij.util.ui.update.MergingUpdateQueue.doFlush(MergingUpdateQueue.java:295)
	at com.intellij.util.ui.update.MergingUpdateQueue.flush(MergingUpdateQueue.java:277)
	at com.intellij.util.ui.update.MergingUpdateQueue.run(MergingUpdateQueue.java:246)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:254)
	at com.intellij.util.Alarm$Request.runSafely(Alarm.java:373)
	at com.intellij.util.Alarm$Request.run(Alarm.java:360)
	at com.intellij.util.concurrency.Propagation.contextAwareCallable$lambda$2(propagation.kt:328)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:272)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:249)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:31)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.executeFirstTaskAndHelpQueue(BoundedTaskExecutor.java:227)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:218)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:215)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.net.NoRouteToHostException: No route to host
	at java.base/sun.nio.ch.Net.connect0(Native Method)
	at java.base/sun.nio.ch.Net.connect(Net.java:579)
	at java.base/sun.nio.ch.Net.connect(Net.java:568)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:593)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:130)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:263)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:183)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.java:224)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.java:108)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.java:88)
	at okhttp3.internal.connection.Transmitter.newExchange(Transmitter.java:169)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:41)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
	at okhttp3.RealCall.execute(RealCall.java:81)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:204)
	at de.mariushoefler.flutterenhancementsuite.utils.PubApi.getPackageLatestVersion(PubApi.kt:86)
	at de.mariushoefler.flutterenhancementsuite.utils.FileParser.mapToVersionDescription(FileParser.kt:61)
	at de.mariushoefler.flutterenhancementsuite.utils.FileParser.getVersionsFromFile(FileParser.kt:21)
	at de.mariushoefler.flutterenhancementsuite.utils.FileParser.checkFile(FileParser.kt:11)
	at de.mariushoefler.flutterenhancementsuite.inspections.PackageUpdateExternalAnnotator.doAnnotate(PackageUpdateExternalAnnotator.kt:26)
	at de.mariushoefler.flutterenhancementsuite.inspections.PackageUpdateExternalAnnotator.doAnnotate(PackageUpdateExternalAnnotator.kt:20)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:207)
	... 37 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@
 
 ### Fixed
 - Fix exception when including a package which is not published on pub.dev (#260)
+- Fix exception when Pub API could not be reached (#261)
 
 ## 1.7.0 - 2023-09-17
 

--- a/src/main/kotlin/de/mariushoefler/flutterenhancementsuite/exceptions/PubExceptions.kt
+++ b/src/main/kotlin/de/mariushoefler/flutterenhancementsuite/exceptions/PubExceptions.kt
@@ -8,6 +8,6 @@ class GetLatestPackageVersionException(p: String) :
 class GetCurrentPackageVersionException(p: String) :
     Exception("Cannot read current version number for package: $p")
 
-class PubApiCouldNotBeReached(p: Exception) : Exception("Pub api could not be reached: $p")
+class PubApiCouldNotBeReachedException(p: Exception) : Exception("Pub api could not be reached: $p")
 
-class PubApiUnknownFormat(p: JsonSyntaxException) : Exception("Unexpected response from Pub api: $p")
+class PubApiUnknownFormatException(p: JsonSyntaxException) : Exception("Unexpected response from Pub api: $p")

--- a/src/main/kotlin/de/mariushoefler/flutterenhancementsuite/utils/FileParser.kt
+++ b/src/main/kotlin/de/mariushoefler/flutterenhancementsuite/utils/FileParser.kt
@@ -2,6 +2,7 @@ package de.mariushoefler.flutterenhancementsuite.utils
 
 import com.intellij.psi.PsiFile
 import de.mariushoefler.flutterenhancementsuite.exceptions.GetLatestPackageVersionException
+import de.mariushoefler.flutterenhancementsuite.exceptions.PubApiCouldNotBeReachedException
 import de.mariushoefler.flutterenhancementsuite.models.VersionDescription
 
 class FileParser(private val file: PsiFile) {
@@ -60,6 +61,8 @@ class FileParser(private val file: PsiFile) {
         val latestVersion = try {
             PubApi.getPackageLatestVersion(dependency.getPubPackageName())
         } catch (e: GetLatestPackageVersionException) {
+            return null
+        } catch (e: PubApiCouldNotBeReachedException) {
             return null
         }
         val currentVersion = dependency.getCurrentPubPackageVersion()

--- a/src/main/kotlin/de/mariushoefler/flutterenhancementsuite/utils/PubApi.kt
+++ b/src/main/kotlin/de/mariushoefler/flutterenhancementsuite/utils/PubApi.kt
@@ -5,8 +5,8 @@ import com.google.gson.JsonSyntaxException
 import com.intellij.openapi.progress.ProgressManager
 import com.intellij.util.io.HttpRequests
 import de.mariushoefler.flutterenhancementsuite.exceptions.GetLatestPackageVersionException
-import de.mariushoefler.flutterenhancementsuite.exceptions.PubApiCouldNotBeReached
-import de.mariushoefler.flutterenhancementsuite.exceptions.PubApiUnknownFormat
+import de.mariushoefler.flutterenhancementsuite.exceptions.PubApiCouldNotBeReachedException
+import de.mariushoefler.flutterenhancementsuite.exceptions.PubApiUnknownFormatException
 import de.mariushoefler.flutterenhancementsuite.models.PubPackage
 import de.mariushoefler.flutterenhancementsuite.models.PubPackageSearch
 import de.mariushoefler.flutterenhancementsuite.models.PubScore
@@ -41,13 +41,13 @@ object PubApi {
                 Gson().fromJson(response, PubPackageSearch::class.java)
             }
         } catch (e: IOException) {
-            throw PubApiCouldNotBeReached(e)
+            throw PubApiCouldNotBeReachedException(e)
         } catch (e: JsonSyntaxException) {
-            throw PubApiUnknownFormat(e)
+            throw PubApiUnknownFormatException(e)
         }
     }
 
-    @Throws(PubApiCouldNotBeReached::class)
+    @Throws(PubApiCouldNotBeReachedException::class)
     fun getPackage(name: String, function: (response: Response<PubPackage>) -> Unit) {
         pubApiService.getPackage(name).enqueue(object : Callback<PubPackage> {
             override fun onResponse(call: Call<PubPackage>, response: Response<PubPackage>) {
@@ -59,12 +59,12 @@ object PubApi {
             }
 
             override fun onFailure(call: Call<PubPackage>, t: Throwable) {
-                throw PubApiCouldNotBeReached(Exception(t))
+                throw PubApiCouldNotBeReachedException(Exception(t))
             }
         })
     }
 
-    @Throws(PubApiCouldNotBeReached::class)
+    @Throws(PubApiCouldNotBeReachedException::class)
     fun getPackageScore(name: String, function: (response: Response<PubScore>) -> Unit) {
         pubApiService.getPackageScore(name).enqueue(object : Callback<PubScore> {
             override fun onResponse(call: Call<PubScore>, response: Response<PubScore>) {
@@ -72,23 +72,27 @@ object PubApi {
             }
 
             override fun onFailure(call: Call<PubScore>, t: Throwable) {
-                throw PubApiCouldNotBeReached(Exception(t))
+                throw PubApiCouldNotBeReachedException(Exception(t))
             }
         })
     }
 
-    @Throws(GetLatestPackageVersionException::class)
+    @Throws(GetLatestPackageVersionException::class, PubApiCouldNotBeReachedException::class)
     fun getPackageLatestVersion(packageName: String): String {
         dependencyCache[packageName]?.let {
             return it
         }
 
         pubApiService.getPackage(packageName).execute().let { response ->
-            response.body()?.let {
-                it.getLatestVersion()?.let { latestVersion ->
-                    dependencyCache[packageName] = latestVersion
-                    return latestVersion
+            if (response.isSuccessful) {
+                response.body()?.let {
+                    it.getLatestVersion()?.let { latestVersion ->
+                        dependencyCache[packageName] = latestVersion
+                        return latestVersion
+                    }
                 }
+            } else {
+                throw PubApiCouldNotBeReachedException(Exception(response.message()))
             }
         }
 

